### ISSUE TITLE
conflicting information about mutations allowed

### ISSUE THREAD

**Issue by petrelharp** [MesserLab/SLiM#179]
Recall that each SLiM mutation may appear "in" more than one tskit mutation, e.g., because of stacking. However, it appears that when reading in tree sequences, there is no check for consistency between these different appearances. For instance, here I construct a tree sequence which has two mutations, each with the same SLiM mutation ID, but different mutation types and different selection coefficients. On loading it into SLiM we find only one mutation, with the second mutation's information:
```
import pyslim, tskit, msprime

def make_tables(ind_ids, mut_types):
    t = tskit.TableCollection(sequence_length=10)
    t.nodes.add_row(time=1.0)
    t = pyslim.annotate_defaults(t.tree_sequence(), model_type='nonWF', slim_generation=1).tables
    pop_md = pyslim.default_slim_metadata('population')
    pop_md['slim_id'] = 0
    t.populations.add_row(metadata=pop_md)
    ind_md = pyslim.default_slim_metadata('individual')
    ind_md['age'] = 4
    ind_md['subpopulation'] = 0
    node_md = pyslim.default_slim_metadata('node')
    for ind_id in ind_ids:
        ind_md['pedigree_id'] = ind_id
        i = t.individuals.add_row(metadata=ind_md, flags=pyslim.INDIVIDUAL_ALIVE, location=[0.0] * 3)
        node_md['slim_id'] = 2 * ind_id
        t.nodes.add_row(time=0.0, population=0, individual=i, flags=1, metadata=node_md)
        node_md['slim_id'] = 2 * ind_id + 1
        t.nodes.add_row(time=0.0, population=0, individual=i, flags=1, metadata=node_md)
    t.sites.add_row(position=0, ancestral_state='')
    for j, mt in enumerate(mut_types):
        mut_md = {'mutation_type': mt, 'selection_coeff': mt + 0.0, 'subpopulation': 0, 'slim_time': -1, 'nucleotide': -1}
        t.mutations.add_row(node=j, site=0, time=1, derived_state='0', metadata={'mutation_list': [mut_md]})
    return t.tree_sequence()

ts = make_tables([0, 1], mut_types=[0, 1])

print(f"The tree sequence has {ts.num_mutations} mutations:")
for mut in ts.mutations():
    for a, md in zip(mut.derived_state.split(","), mut.metadata['mutation_list']):
        print(f" SLiM ID {a}: type={md['mutation_type']}, selection_coeff={md['selection_coeff']}")

ts.dump("dup_pops.trees")
```
produces:
```
The tree sequence has 2 mutations:
 SLiM ID 0: type=0, selection_coeff=0.0
 SLiM ID 0: type=1, selection_coeff=1.0
```
and running
```
initialize() {
	initializeSLiMModelType("nonWF");
	initializeSLiMOptions(keepPedigrees=T);
	initializeTreeSeq();
	initializeMutationType("m0", 0.5, "e", 1/100);
	initializeMutationType("m1", 0.5, "e", 1/100);
	initializeGenomicElementType("g1", m0, 1);
	initializeGenomicElement(g1, 0, 9);
	initializeMutationRate(1e-10);
	initializeRecombinationRate(0.0);
}

1 early() {
    sim.readFromPopulationFile("dup_pops.trees");
    for (pop in sim.subpopulations)
        catn("pop" + pop.id + " ids: " + paste(pop.individuals.pedigreeID));
    for (mut in sim.mutations)
        catn("mut" + mut.id + ": type=" + mut.mutationType.id + " selectionCoeff=" + mut.selectionCoeff);
}   

3 early() { sim.simulationFinished(); }
```
produces
```
pop0 ids: 0 1 2 3
mut0: type=1 selectionCoeff=1
```
(note however that the correct genomes do carry the mutation).

**Comment by petrelharp** [comment]
Note: for some reason, most of my attempts to produce this bug resulted in an error like
```
ERROR (SLiMSim::CrosscheckTreeSeqIntegrity): (internal error) genome/allele size mismatch at position 26: the treeseq has 1 mutation of mutid 0, SLiM has 0 segregating and 0 fixed mutation(s).
```
I don't know what makes this example avoid the crosscheck.

**Comment by bhaller** [comment]
Hmm.  Yes, it would be nice to catch these errors; but this is pretty pathological, right?  I mean, it should probably produce an error to try to create a malformed file this this in the first place.  It's a good goal to be robust to any kind of corrupted file that we're asked to load, in principle, but it feels pretty unlikely that any end user would actually hit this...?

**Comment by petrelharp** [comment]
It's an easy mistake to make if you add SLiM mutations with more than one call to msprime (e.g., first add deleterious mutations; then add positive mutations) - you need to remember to tell it to offset the mutation IDs by the number already there.  We ran into this because we generated initial diversity for two species separately with msprime, and merged the tree sequences together, but forgot to make the mutation IDs unique across the two tree sequences.

**Comment by bhaller** [comment]
I see.  OK, if it has bitten a real-world case then that makes it more urgent to fix.  Thanks.

**Comment by bhaller** [comment]
Hey @petrelharp.  So, we have a comment in SLiMSim::__TabulateMutationsFromTables (slim_sim.cpp:8127) that says:

`// tabulate the mutations referenced by this entry, overwriting previous tabulations (last state wins)
`

That makes it look to me like this behavior is deliberate.  Contemplating why that might be, I have realized that in fact it makes sense in general.  The problem is that Mutation objects in SLiM are mutable; you are allowed to change the selection coefficient, the mutation type, etc. of a mutation while a simulation is running.  The information recorded in a given derived state record just records the state of the mutation at the point in time that that snapshot was made.  If the mutation's state is changed later, a later snapshot will (hopefully) record that (I suspect we have a bug there, that a simulation could change a mutation's state but if that mutation does not happen to get snapped in a new derived state before saving, the final mutation state will not be correctly recorded in the tree sequence).  So when reading in a tree sequence, we allow differences, and just take the last state recorded.  So I think this is not a bug; it is necessary to support supported behavior in scripts.

It does sound like maybe there's a bug in the msprime stuff (?), if it allows you to overlay new mutations that use the same mutation IDs as previously existing mutations, or if a merge doesn't do the right thing with that, right?

If you agree that this is not a bug, feel free to close it.

**Comment by petrelharp** [comment]
Ah, I see.  Well, then.

I don't think it's the job of msprime to look out for this, as it's very SLiM-specific (and not even necessarily a bug). We should figure out how to make it easy to do the right thing in pyslim (maybe just by providing a `next_mutation_id` method or something).

So, I guess it's Not A Bug. Although it's an easy mistake to make, so I think there should be a warning, only on reading in the tree sequence?

> (I suspect we have a bug there, that a simulation could change a mutation's state but if that mutation does not happen to get snapped in a new derived state before saving, the final mutation state will not be correctly recorded in the tree sequence). 

Ah, hm. I'm tempted to say that if you're doing this sort of thing, it's up to you to figure out how to write out the information and adjust it properly when reading back in the tree sequence.I mean, if a mutation has a vector of effects (on several traits, for instance), set with setValue( ), that won't get recorded either.

**Comment by bhaller** [comment]
> So, I guess it's Not A Bug. Although it's an easy mistake to make, so I think there should be a warning, only on reading in the tree sequence?

Hmm.  Well, I'm not sure it even merits a warning – I mean, tweaking mutations during a run is the standard, recommended way to do certain things.  So then it should probably not produce a warning – warnings should be reserved for cases where it is at least "pretty likely" that there is a problem, right?  If it's an easy mistake to make when overlaying mutations or whatever when using msprime/pyslim, then it should probably produce a warning – or even an error! – in *that* workflow, since in that workflow it is, indeed, very likely to represent a mistake.  But I don't think SLiM should warn when seeing this on read, because it can easily/often result for standard, accepted practice.

**Comment by bhaller** [comment]
I think this is Not A Bug, at least on the SLiM side.  Closing; feel free to reopen if you disagree.

**Comment by petrelharp** [comment]
Well, this is unfortunate, because we cannot raise an error or warning for this on the msprime side. I guess I should aim to provide a pyslim wrapper to mutate( ) that includes this check.

Oh, wait - there is *one* check for consistency that we *can* (and should) do - that the *positions* along the sequence of mutations with matching IDs are the same. That can't be changed (right?) and would catch out the bug.

I can't re-open this, though - you've got to do that.

**Comment by bhaller** [comment]
Yes, true – positions should not change.  Perhaps there's another property or two that should not change, too.  OK, we can check for that in SLiM.
